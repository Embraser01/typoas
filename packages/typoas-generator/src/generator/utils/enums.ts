import { SchemaObject } from 'openapi3-ts';
import { factory, EnumMember } from 'typescript';
import { screamingSnakeCase } from './operation-name';

export function canConvertSchemaToEnum(schema: SchemaObject): boolean {
  return (
    schema.type === 'string' &&
    schema.enum !== undefined &&
    schema.enum.every((e) => ![null, '', undefined].includes(e)) &&
    !schema.nullable
  );
}

export function createEnumMembersFromSchema(
  schema: SchemaObject,
): EnumMember[] {
  if (canConvertSchemaToEnum(schema)) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return schema.enum!.map((e) => {
      return factory.createEnumMember(
        factory.createIdentifier(screamingSnakeCase(e)),
        factory.createStringLiteral(e, true),
      );
    });
  }
  throw new Error('Cannot create enum members from this schema');
}
