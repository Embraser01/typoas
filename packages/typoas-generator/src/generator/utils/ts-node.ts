import {
  createPrinter,
  EmitHint,
  factory,
  NewLineKind,
  Node,
  NodeFlags,
  SyntaxKind,
} from 'typescript';

export const getStringFromNode = (node: Node): string => {
  const printer = createPrinter({ newLine: NewLineKind.LineFeed });
  const resultFile = factory.createSourceFile(
    [],
    factory.createToken(SyntaxKind.EndOfFileToken),
    NodeFlags.Const,
  );
  return printer.printNode(EmitHint.Unspecified, node, resultFile);
};
