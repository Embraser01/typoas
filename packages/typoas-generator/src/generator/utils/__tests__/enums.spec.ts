import { describe, expect, it } from '@jest/globals';
import { SchemaObject } from 'openapi3-ts';
import { factory, SyntaxKind } from 'typescript';
import { canConvertSchemaToEnum, createEnumMembersFromSchema } from '../enums';
import { getStringFromNode } from '../ts-node';

describe('canConvertSchemaToEnum', () => {
  it('should handle string schema', () => {
    const schema: SchemaObject = { type: 'string', enum: ['a', 'b', 'c'] };
    expect(canConvertSchemaToEnum(schema)).toBe(true);
  });

  it('should not handle string schema without enum', () => {
    const schema: SchemaObject = { type: 'string' };
    expect(canConvertSchemaToEnum(schema)).toBe(false);
  });

  it('should not handle non string schemas', () => {
    const schema: SchemaObject = { type: 'number', enum: [2, 4, 7] };
    expect(canConvertSchemaToEnum(schema)).toBe(false);
  });

  it('should not handle enums with null or empty values', () => {
    const schema: SchemaObject = { type: 'string', enum: [null, 'test'] };
    expect(canConvertSchemaToEnum(schema)).toBe(false);
  });

  it('should not handle nullable enums', () => {
    const schema: SchemaObject = {
      type: 'string',
      enum: ['a', 'test'],
      nullable: true,
    };
    expect(canConvertSchemaToEnum(schema)).toBe(false);
  });
});

describe('createEnumMembersFromSchema', () => {
  it('should handle string schema', () => {
    const schema: SchemaObject = {
      type: 'string',
      enum: ['dog-and-cat', 'dog', 'catWithoutDog'],
    };
    const node = factory.createEnumDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      factory.createIdentifier('Test'),
      createEnumMembersFromSchema(schema),
    );
    expect(getStringFromNode(node)).toMatchSnapshot();
  });

  it('should handle enum with special chars in it', () => {
    const schema: SchemaObject = {
      type: 'string',
      enum: ['my:complex:enum', 'and:this/one'],
    };
    const node = factory.createEnumDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      factory.createIdentifier('Test'),
      createEnumMembersFromSchema(schema),
    );
    expect(getStringFromNode(node)).toMatchSnapshot();
  });
});
