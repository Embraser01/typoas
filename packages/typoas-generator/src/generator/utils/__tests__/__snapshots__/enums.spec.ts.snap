// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createEnumMembersFromSchema should handle enum with special chars in it 1`] = `
"export enum Test {
    MY_COMPLEX_ENUM = 'my:complex:enum',
    AND_THIS_ONE = 'and:this/one'
}"
`;

exports[`createEnumMembersFromSchema should handle string schema 1`] = `
"export enum Test {
    DOG_AND_CAT = 'dog-and-cat',
    DOG = 'dog',
    CAT_WITHOUT_DOG = 'catWithoutDog'
}"
`;
